### PR TITLE
Add ResultImageListener to intercept destination channel rendering

### PR DESCRIPTION
**Pull request by tribal-tec**

### CODE REVIEW

**Comment by tribal-tec** [comment]
FYI @chevtche @favreau

**Comment by tribal-tec** [comment]
Retest this please

File: eq/client/channel.cpp [hunk 1]
```diff
@@ -88,7 +88,7 @@ using detail::STATE_FAILED;
 
 Channel::Channel( Window* parent )
         : Super( parent )
-        , _impl( new detail::Channel )
+        , _impl( new detail::Channel( *this ))
```
Comment (eile) [inline eq/client/channel.cpp]: This typically gives a warning 'this used in initializer list'

Comment (tribal-tec) [inline eq/client/channel.cpp]: pragma push then? It's MSVC only, right?

File: eq/client/dc/proxy.cpp [hunk 2]
```diff
@@ -89,38 +89,36 @@ class Proxy : public boost::noncopyable
...
-        const size_t newSize = pvp.w * pvp.h * bytesPerPixel;
-        buffer.resize( newSize );
-
-        // OPT: use FBO texture directly to download
-        if( !_fboDownload( ))
-            _textureDownload();
+        // copy pixels to perform swapYAxis()
+        const size_t dataSize = image.getPixelDataSize( Frame::BUFFER_COLOR );
+        buffer.resize( dataSize );
+        memcpy( buffer.getData(), image.getPixelPointer( Frame::BUFFER_COLOR ),
```
Comment (eile) [inline eq/client/dc/proxy.cpp]: Implement buffer.set( ptr, size )?

Comment (tribal-tec) [inline eq/client/dc/proxy.cpp]: True. Looking at the buffer API, this one seems missing. Will do.

Comment (tribal-tec) [inline eq/client/dc/proxy.cpp]: Actually, buffer.replace() already exists and does what we want here.

File: eq/client/detail/channel.ipp [hunk 3]
```diff
@@ -41,45 +46,94 @@ enum State
...
         {
-            lunchbox::RNG rng;
-            color.r() = rng.get< uint8_t >();
-            color.g() = rng.get< uint8_t >();
-            color.b() = rng.get< uint8_t >();
+            ResultImageListeners::iterator i =
+                    std::find( resultImageListeners.begin(),
+                               resultImageListeners.end(), &frameWriter );
+            if( i == resultImageListeners.end( ))
+                addResultImageListener( &frameWriter );
```
Comment (eile) [inline eq/client/detail/channel.ipp]: move to configInit/Exit?

File: eq/client/detail/fileFrameWriter.h [hunk 4]
```diff
@@ -27,19 +27,20 @@ namespace detail
...
     FileFrameWriter();
     ~FileFrameWriter();
 
-    void write( eq::Channel* channel );
+    void notifyNewImage( const eq::Channel& channel,
+                         const eq::Image& image ) final;
 
 private:
-    eq::Image _image;
+    void _write( eq::Channel* channel );
```
Comment (eile) [inline eq/client/detail/fileFrameWriter.h]: Not implemented.

File: eq/client/resultImageListener.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+#define EQ_RESULTIMAGELISTENER_H
+
+#include <eq/client/types.h>
+
+namespace eq
+{
+
+/**
+ * The result image listener gets notified on new images produced by
+ * destination channels.
```
Comment (eile) [inline eq/client/resultImageListener.h]: @sa Channel::addResultImageListener.

File: eq/client/channel.cpp (hunk 1 above)
Comment (eile) [inline eq/client/channel.cpp]: iirc yes - but you don't seem to need it. frameViewFinish() already hast eq::Channel* as param.

File: eq/client/detail/channel.ipp (hunk 3 above)
Comment (tribal-tec) [inline eq/client/detail/channel.ipp]: I thought about it too, but the way is initially designed is that one could in theory add and remove the writer at runtime by changing the channel's attribute. If this is not desired or a use case for now, I'm happy to move this to configInit/Exit.

File: eq/client/detail/fileFrameWriter.h (hunk 4 above)
Comment (tribal-tec) [inline eq/client/detail/fileFrameWriter.h]: Good catch, will be removed.

File: eq/client/resultImageListener.h (hunk 5 above)
Comment (tribal-tec) [inline eq/client/resultImageListener.h]: Done

File: eq/client/channel.cpp (hunk 1 above)
Comment (tribal-tec) [inline eq/client/channel.cpp]: Good point. Removed.

**Comment by tribal-tec** [comment]
Retest this please

**Comment by eile** [comment]
+1

**Pull request merged**